---
source: Rmd
title: "Hands On: Performing Enrichment Analysis on Differentially Expressed Genes" 
exercises: 45
objectives:
- Gain functional insight from a list of differetially expressed genes
- Perform a GO enrichment analysis
- Simplify the results of the GO enrichment analysis
---

```{r, include=FALSE}
source("../bin/chunk-options.R")
knitr::fig_path(options = list(fig.path = "../fig/"))
```

```{r, message=F, warning=F, include = FALSE}
#BiocManager::install("clusterProfiler", version = "3.8")
#BiocManager::install("pathview")
#BiocManager::install("enrichplot")
library(clusterProfiler)
library(enrichplot)
# we use ggplot2 to add x axis labels (ex: ridgeplot)
library(ggplot2)
```

## Introduction to GO Term Enrichment Analysis

When we have a large list of genes of interest, such as a list of differentially expressed genes obtained from an RNA-Seq experiment, how do we extract biological meaning from it?

One way to do so is to perform functional enrichment analysis. This method consists of applying statistical tests to verify if genes of interest are more often associated to certain biological functions than what would be expected in a random set of genes. In this tutorial you will learn about enrichment analysis and how to perform it.

## What is the Gene Ontology?

The Gene Ontology (GO) is a structured, controlled vocabulary for the classification of gene function at the molecular and cellular level. It is divided in three separate sub-ontologies or GO types: biological process (e.g., signal transduction), molecular function (e.g., ATPase activity) and cellular component (e.g., ribosome). These sub-ontologies are structured as directed acyclic graphs (a hierarchy with multi-parenting) of GO terms.

This week, we are going to focus on the **Biological Function** GO type. Here is an example of the hierarchy that would get you from a relatively specific biological processs, **`negative regulation of programmed cell death`** all the way back up to the general **`biological_process`** term. You can see that terms can be related to one another by a variety of different relationships, such as **Positive or negative regulation** or being **Part of** another term.

![Example GO term hierarchy and list of relationships](../fig/GO_0043069_goAncestry.png)

## What actually happens when we perform functional enrichment?

### First, we gather the data we need:

-   A set of genes of interest (e.g., differentially expressed genes): Our differentially expressed genes

-   A set with all the genes to consider in the analysis: population set (which must contain the study set): In this case, all genes in the mouse genome

-   GO annotations, associating the genes in the population set to GO terms

-   The GO ontology, with the description of GO terms and their relationships

### Then, we plan calculate enrichment values:

For each GO term, we need to count the frequency (k) of genes in the study set (n) that are associated to the term, and the frequency (K) of genes in the population set (N) that are associated to the same term. Then we test how likely would it be to obtain at least k genes associated to the term if n genes would be randomly sampled from the population, given the frequency K and size N of the population. **That is, how unlikely is it that a particular GO Term would occur so frequently in our sample of differentially expressed genes just by chance, compared to the background.**

**The appropriate statistical test is the one-tailed variant of Fisher's exact test, also known as the hypergeometric test for over-representation.** When the one-tailed version is applied, this test will compute the probability of observing at least the sample frequency, given the population frequency. The hypergeometric distribution measures precisely the probability of k successes in n draws, without replacement, from a finite population of size N that contains exactly K successful objects:

![Formula for hypergeometric formula assuming variable names as above](../fig/goenrichment_formula.png)

## Hands-On: Conducting a GO Term Enrichment Analysis

We already have our `pregnant_lactate_limma` object loaded into RStudio, so we aleady have all of the information from our DE analysis that we need. We need to re-arrange this data a bit so it is in a `named vector` of just the `Log fold change`, each value named after its corresponding `EntrezID` and sorted in decreasing order by the fold change as required by the **`clusterProfiler`** package we will be using.

```{r, echo=FALSE}
 pregnant_lactate_limma <- read.csv("../data/limma-voom_basalpregnant-basallactate.csv",header=TRUE)
```

### Preparing our DE input

```{r eval=TRUE}
# we want the log2 fold change stored as a vector
original_gene_list <- pregnant_lactate_limma$logFC

# name the vector with the EntrezIDs
names(original_gene_list) <-pregnant_lactate_limma$ENTREZID

# omit any NA values (here shouldn't be any missing values for either variable, but this is always a good thing to check.) 
gene_list<-na.omit(original_gene_list)

# sort the list in decreasing order (required for clusterProfiler)
gene_list = sort(gene_list, decreasing = TRUE)
```

You should now have a named vector called `gene_list` in your environment.

> ## How many items are in the `gene_list` object? What do each of them represent?
>
> `14187 items`. Each of these represent the `log(2)fold` change of one differentially expressed gene, and the corresponding `EntrezID`.
>
> You can look at the `gene_list` item under **Values** in your **Environment Pane** to find this information.
>
> Alternatively, you can run either of the following commands to get the same information.
>
> ```{r eval=TRUE}
> length(gene_list)
> str(gene_list)
> ```
>
{: .solution}

### Loading mouse genome GO annotations

I'm using data generated from differential expression in mice, so I install and load the annotation "org.Mm.eg.db" below, which contains all of the GO terms available for genes in the mouse genome. If you want to work on a different organism, there are ready-to-go GO annotation sets for 19 of the most commonly used model organisms. Look up the code for your particular organism here: <http://bioconductor.org/packages/release/BiocViews.html#___OrgDb>

**Because of the large file size, I have already installed the mouse GO annotation package for you using the commands below:**

```{r,  message=F, warning=F, cache=TRUE}
# SET THE DESIRED ORGANISM HERE
organism = "org.Mm.eg.db"
# Install annotation set like a Bioconductor package
BiocManager::install(organism, character.only = TRUE)

```

**You just need to run the following command to load the annotation package into your environment**

```{r, message=F, warning=F}
library(organism, character.only = TRUE)
```

### Gene Set Enrichment

We are going to be running the following command to do a gene set enrichment analysis:

```{r, eval=FALSE}
gse <- gseGO(geneList=gene_list, 
             ont ="BP", 
             keyType = "ENTREZID",
             minGSSize = 3, 
             maxGSSize = 800, 
             pvalueCutoff = 0.05, 
             verbose = TRUE, 
             OrgDb = organism, 
             pAdjustMethod = "none")
```

Let's see what each of these parameters do:

-   **geneList:** Where are we storing our list of genes that we created above?

-   **ont:** Choose one of the three GO term sub-ontologies to get GO terms from (BP = biological process, MF = molecular function, CC = cellular component, ALL = all three).

-   **keyType:** This is how our genes are identified in our data set**.** The options vary for each organism's annotation, but can include `ENTREZID`, `REFSEQ`, `ENSEMBL` IDs among many other possibilities. This tells this function how to connect the ID numbers to GO terms.

-   **minGSSize:** minimum size of each geneSet for analyzing. For example, `minGSSize = 3` tells gsea() to set aside any GO terms that have three or fewer genes annotated with that term. GO terms with few genes generally do not contribute much to enrichment analysis.

-   **maxGSSize:** Maximum size of each geneSet for analyzing. For example, `maxGSSize = 800` tells gsea() to set aside any GO terms more than 800 genes annotated with that term. GO terms with many genes tend to be very general, non-specific GO terms which would not be particularly interesting to us biologically.

-   **pvalueCutoff:** Significance threshhold for deciding whether a GO term is significantly enriched in our data set compared with the background of all mouse genes.

-   **verbose:** Should this function report back lots of info as it runs in case we need to troubleshoot?

-   **OrgDb:** Where are we storing the annotation info for the organism of our choice?

-   **pAdjustMethod:** How are we adjusting our p-value for multiple testing or false discovery rate? Discussion of these methods is outside the scope of this course.

**Okay, let's actually run this command!** You will probably see a lot of info on your screen and it may take a moment to run. You can ignore any warnings for now!

```{r, cache=TRUE}
gse <- gseGO(geneList=gene_list, 
             ont ="BP", 
             keyType = "ENTREZID",
             minGSSize = 3, 
             maxGSSize = 800, 
             pvalueCutoff = 0.05, 
             verbose = TRUE, 
             OrgDb = organism, 
             pAdjustMethod = "none")
```

> ## Workaround if gseGO() stalls 
>  I have run gseGO() on this tutorial data set myself and exported it such that you can upload `gse` object directly into R if the **gseGO()** function is taking forever.  
> 
> 1. Download this file and upload it to RStudio Cloud: [gse_tutorial.RData](../files/gse_tutorial.RData). 
> 2. Upload this file into RStudio as you did for the Volcano Plot script or any other file you need to get into RStudio. 
> 3. Then, just click on the name of the file in your "Files" pane. 
> 4. R should automatically run a load() function, and the object "gse" should appear in your environment!
> 
{: .callout} 

## Examining GO term enrichment results

### Dataframe of results

As long as you now have a large (99 MB) object called `gse` in your environment, you are good to go! There are two really interesting objects that are part of the big list of results stored in `gse`. The first is a `data frame` containing the results:

```{r}
# Save the results dataframe as its own object so we can manipulate it. Use the "@" operator to select this part of the whole list. 
pregnant_lactate_gsea_BP <- gse@result
str(pregnant_lactate_gsea_BP)
```

There is a lot going on in the `pregnant_lactate_gsea_BP` data frame we just extracted, but here are some highlights. There is one row for each significantly enriched biological process GO term from our Gene Enrichment analysis.

-   **ID:** The official GO term ID

-   **Description:** Brief description of the biological process associated with a GO Term ID

-   **setSize:** The number of genes annotated with a particular GO term. One gene can potentially be annotated with many GO terms.

-   **p-value:** Result of testing for whether a GO term is significantly enriched in our data set compared with the background of all mouse genes.

-   **NES**: Normalized Enrichment Score. The primary result of the gene set enrichment analysis is the enrichment score (ES), which reflects the degree to which a gene set is overrepresented at the top or bottom of a ranked list of genes. The sign on the score simply indicates when end of your ranked gene list is enriched. You provide the rank list of genes, so the biological interpretation is up to you. **Since our genes were ranked by FoldChange, then the positive scores are associated with upregulated genes and negative scores are associated with downregulated genes (with respect to the `pregnant` samples in this example). By normalizing the enrichment score, GSEA accounts for differences in gene set size and in correlations between gene sets and the expression dataset, making NES values more comparable across data sets**

> ## What is the average number of genes per GO term?
>
> Think back to one of our old favorite basic R functions, mean().
>
> ```{r eval=TRUE}
> mean(pregnant_lactate_gsea_BP$setSize)
> ```
>
> This means (hah) that there are on average about 108 genes that were annotated with each GO term. 
{: .solution}

### Lists of genes per GO Term

The `geneSets` item in the `gse` results will likely be something useful if you were interested in knowing more about the genes that were annotated with a particular GO term. Looking at its structure, we can see that `geneSets` is a list of vectors, one per GO term. **Each vector contains the EntrezIDs of the genes annotated with that GO term, for each GO term in the analysis whether it was significantly enriched or not.**

```{r}
str(gse@geneSets)
```

Let's say that we wanted to know which genes were associated with the <GO:0000085>, which corresponds to the **mitotic G2 phase** biological process:

```{r}
gse@geneSets$`GO:0000085`
```

As we can see, there is only one gene in there (EntrezID = 50883). Looking up this ID on NCBI we can see that this corresponds to the `**Chek2 checkpoint kinase 2**`, which makes a lot of sense given the biological process GO term above.

## Visualizing GO Term Enrichment Results

The `**clusterProfiler**` package has a variety of options for viewing the over-represented GO terms. I will go through a few typical types of plots, although there are many other possibilities for you to explore on your own.

### Dotplots

The following command will produce a **dotplot**

```{r, eval=FALSE}
dotplot(gse, showCategory=10)
```

![Dotplot using our enrichment results, showing top 10 GO categories and otherwise using default parameters](../fig/Dotplot_GeneRatio_10categories.png)
*Dotplot using our enrichment results, showing top 10 GO categories and otherwise using default parameters*

The dotplot shows the number of genes associated with the Top 10 most enriched GO terms and adjusted p-values for these terms (color). The terms are ordered on the X-axis by the `**GeneRatio**` variable by default, which is simply the `count/setSize` for a GO term, where the `count` is the number of genes in a GeneSet (i.e. annotated with that particular GO term) and the `setSize` is the number of genes in the background mouse annotation dataset. Basically, another way of visualizing group size.

I have adjusted the plot dotpot so that it now shows the top 20 GO terms and has the Normalized Enrichment Score on the x-axis, to help us understand whether these GO terms are generally upregulated or downregulated in the `basalpregnant vs. basallactate` comparison.

![The dotplot shows the number of genes associated with the first 20 GO terms (by size) and the p-adjusted values for these terms (color). The terms are ordered on the X-axis by the \*\*Normalized Enrichment Score\*\* variable](../fig/dotplot_20category.png) *The dotplot shows the number of genes associated with the first 20 GO terms (by size) and the p-adjusted values for these terms (color). The terms are ordered on the X-axis by the `**Normalized Enrichment Score**` variable*

> ## What command produced this version of the dotplot?
>
> First, we adjust the `showCategory` parameter to be 20 instead of 10.
>
> Second, you will need to use the `help("dotplot")` command to see all of the options for this command to see which parameter changes the x-axis.
>
> Putting this all together and running `dotplot()` again:
>
> ```{r eval=FALSE}
> dotplot(gse, showCategory=20,x="NES")
> ```
>
> Try running this on your own to make sure it works! 
{: .solution} 

We may also want to be able to modify this plot further, to separate out which top GO Term categories are relative activated (upregulated - positive NES) or suppressed (downregulated - negative NES). We can do exactly this by telling `**dotplot()**` to split tne NES values by their `sign` , and then tell R to plot each sign in a separate `facet`, or panel.

```{r, eval=FALSE}
#Make sure to include the periods!
dotplot(gse, showCategory=10,x="NES", split=".sign") + facet_grid(.~.sign)
```

![Two-panel dotplot showing us which of the Top 20 largest GO Term groups are generally upregulated (positive NES, left panel), or generally downregulated (negative NES, right panel)](../fig/Dotplot_Top10_BySign.png)
*Two-panel dotplot showing us which of the Top 20 largest GO Term groups are generally upregulated (positive NES, left panel), or generally downregulated (negative NES, right panel.*


### Enrichment Maps

Enrichment maps are a great way to explore which GO terms might form functional clusters of related genes. An enrichment map organizes significanly GO enriched terms into a network with edges connecting overlapping gene sets. In this way, mutually overlapping gene sets tend to cluster together, making it easy to identify potential functional modules. We can generate an enrichment map of the top GO terms from our data set in two steps: 

```{r, eval=FALSE}
# First, generate a matrix of pairwise similarities between GO terms based on the overlap between their lists of genes
gse1 <- pairwise_termsim(gse)

# We can extract the data frame that contains the calculated pairwise distances between the GO terms 
termsim_table <- gse1@termsim
str(termsim_table)
```



```{r, eval=FALSE}
#Then, simply use the emapplot() function to create a network of overlaps between signifiant GO terms: 
emapplot(gse1)
```

![Enrichment map of the top significantly enriched GO terms amongst the differentially expressed genes from our basalpregnant vs. pregnantlactate comparison.](../fig/EnrichmentMap_DefaultLayout.png)
*Enrichment map of the top significantly enriched GO terms amongst the differentially expressed genes from our basalpregnant vs. pregnantlactate comparison.*

**Interpreting the enrichment map**: In the background, the emapplot() function has a minimum similarity threshold for whether two terms are connected. The default value of this thresh is 0.2, for a value that ranges from 0.0 (no overlapping genes) to 1.0 (complete overlap). The default layout algorithm causes gene sets with high overlap to cluster closely together. **Intuitively, this means that tight clusters of GO terms have high amounts of overlap between them, and vice versa. One thing we notice is that some GO terms are not linked to any other, or very few, other terms.** 

Here is a screenshot of a portion of the Pairwise Similarity Matrix. We can compare what we notice about the `inflammatory response` in the Enrichment Map, and what we see in the table in the similarity table. 

![Portion of the table of pairwise similarities between GO terms](../fig/pairwise_GOterm_similarities.png)
*Portion of the table of pairwise similarities between GO terms. Looking at the "inflammatory response" column, we can see that all of the values are very small (less than the 0.2 thresshold). This is played out in the enrichment plot, in which that GO term is not connected to any other terms*. 

**One important parameter you can change is the `layout` parameter, which allows you to choose how you would like mapping results to be displayed.** Just for fun, let's try modifying that parameter to `star`, which should be fairly self-explanatory once you see those results!
```{r, eval=FALSE}
# We can extract the data frame that contains the calculated pairwise distances between the GO terms 
emapplot(gse1,layout="star")
```

![Enrichment map of the top significantly enriched GO terms amongst the differentially expressed genes from our basalpregnant vs. pregnantlactate comparison, with a `star` layout.](../fig/EnrichmentMap_StarLayout.png)
*Enrichment map of the top significantly enriched GO terms amongst the differentially expressed genes from our basalpregnant vs. pregnantlactate comparison, with a `star` layout. Notice that the "inflammatory response" GO term is still off by itself on the lefthand side of the plot*

### Ridgeline Plots

Ridgeline plots help combine our information about our significantly enriched GO terms and the actual logfold changes reported for the differentially expressed genes within that group. 

```{r, eval=FALSE}
ridgeplot(gse1)
```


![Example ridgeline plot](../fig/Ridgeline_Plot_Example.png)